### ISSUE TITLE
Implement batch operations and caching for gradient embeddings

### ISSUE THREAD

**Issue by Nikhil-Prasad** [zou-group/textgrad#185]
## Description
Optimize the embedding and retrieval pipeline for production use.

### 1. Batch Embedding
```python
# Current (slow)
for gradient in gradients:
    embedding = model.encode(gradient)
    store.add(gradient, embedding)

# Optimized (fast)
embeddings = model.encode(gradients, batch_size=32)
store.add_batch(gradients, embeddings)
```

### 2. Embedding Cache
- LRU cache for frequently accessed gradients
- Precompute embeddings for common error patterns
- Cache warming on startup

### 3. Async Operations
```python
# Non-blocking storage
async def store_gradient_async(gradient, metadata):
    embedding = await embed_async(gradient)
    await store.add_async(gradient, embedding, metadata)
```

### 4. Performance Monitoring
- Embedding latency percentiles
- Cache hit rates
- Storage operation timing
- Memory usage tracking

### 5. Query Optimization
- Approximate nearest neighbor search
- Metadata pre-filtering
- Result caching for repeated queries

## Expected Improvements
- 10x faster batch operations
- <10ms retrieval latency (p99)
- 50% reduction in embedding API calls
- Minimal impact on training loop speed

## Benchmarks to Track
- Gradients processed per second
- End-to-end training time comparison
- Memory overhead
- Cost reduction (for paid embedding APIs)

**Comment by Nikhil-Prasad** [comment]
Apologies accidentally created issues on this repo via gh client.
